mlflow/store/tracking/sqlalchemy_store.py

```diff
@@ -222,6 +223,29 @@ def _get_or_create(self, session, model, **kwargs):
 
         return instance, created
 
+    def _get_or_create_many(self, session, model, list_of_args):
```

Comment:
Re-examining the `_get_or_create` code after a few years, I'm wondering if it's truly necessary to have this logic for point writes or for batch writes. The only place where the `created` boolean output of `_get_or_create` is used is https://github.com/mlflow/mlflow/blob/78c8dfad74dbfa06b9707054aeee44be4398bf12/mlflow/store/tracking/sqlalchemy_store.py#L591, where we check it to determine whether or not to update `latest_metrics`.

In all other cases, `_get_or_create` is duplicating the function of an integrity constraint.

Can we remove the `_get_or_create_many` semantics from this change and instead in favor of implementing the the following logic for the following procedures?

**log_params**

1. Create a session
2. Within the session, perform a `_get_run()` operation to fetch all of the params associated with the run
3. Attempt to insert the batch of parameters into the `params` table via `session.add_all()` & `session.commit()`
4. In case of an integrity error, check to see if there are any value differences between the new batch (from 3) and the existing run parameters (from 2). If there are, throw a specific MlflowException. If there aren't, return successfully in order to provide idempotency (`Params: once written, param values cannot be changed (attempting to overwrite a param value will result in an error). However, logging the same param (key, value) is permitted - that is, logging a param is idempotent.` from https://mlflow.org/docs/latest/rest-api.html#log-batch).

This is very similar to https://github.com/mlflow/mlflow/blob/38d9f0efc5d6a6d4ef2b591e95210e0dc3bc5296/mlflow/store/tracking/sqlalchemy_store.py#L757-L767, except we don't `reraise` in the case where there isn't any mismatch between the new parameter values and the existing values in the DB.

**log_metrics**

1. Create a session
2. Attempt to insert the batch of metrics into the `metrics` table via `session.add_all()` & `session.commit()`
3. Unconditionally attempt to update the `latest_metrics` table
4. If any integrity errors are encountered, raise them directly to the caller

**set_tags**

1. Create a session
2. Attempt to upsert the batch of tags into the `tags` table via `session.merge()`
3. If any integrity errors are encountered (unlikely), raise them directly to the caller

Reply:
After implementing these changes, I'd be curious to see whether the performance is even better, given that we'd no longer perform entity lookups prior to insertion.

Reply:
Hi @dbczumar , thank you so much for such detailed comment. 

I didn't want to make radical changes and keep things behave as they do in logging one by one, but I myself think these parts should be re-designed as well. I totally agree with your comment on the necessity of `_get_or_create`.

**log_params**

Yeah, it makes sense. You expect to ignore the integrity error if the parameters are same. And even if one parameter does not match, we should rollback the changes. Did I understand correctly?

**log_metrics**

Makes sense. And I should update latest_metrics table in batch afterwards.

**set_tags**

Makes sense. I also would not expect an integrity error since we merge.

Reply:
I tried to implement your requests but there is some problems:

sqlalchemy does not support batch upsert operations via session.merge(). so I didn't implement it.
since we change the behavior in _log_params, now some test cases fail. The problem is that we handle the exception (and conditionally raise) for the whole batch, and during handling it, the whole session is rolled back. But the test cases ([test_log_batch_param_overwrite_disallowed_single_req](https://github.com/mlflow/mlflow/blob/2bb754ace70fd72f99a60f6e9bca146f4dba5071/tests/store/tracking/test_sqlalchemy_store.py#L1615)) still expect that one of the params is logged at all..